### DISCUSSION TITLE
Separate layouts into a separate method

### DISCUSSION THREAD

**Discussion by MrRymen** [eclipse/windowbuilder#177] [Ideas]
Dear WindowBuilder developers.

When we use GroupLayouts in WindowBuilder, it makes a huge code mess that is hard to read. It is especially difficult when there are tons of components, such as JPanels and Tabs, etc.

**I suggest adding an option in:**
**Windows -> Preferences -> WindowBuilder -> Swing -> Code Generation**
That will be called: **"Generate layouts into a separate method from the rest of the code"**

When this option will be turned on, the window builder should generate any layout managers (such as GroupLayout and others) into a separate method called: layoutInitializer().

Then method **layoutInitializer()** will be called inside the constructor of the JFrame same as initialize().

Of course, in such scenario most of the JPanels and other components somehow related to GroupLayout will have to be field as well as the GroupLayout itself in order for this to work.

**There are several benefits to this;**
It will be easier to read the code.
It will save time for all developers, especially for new developers.
It will look nicer and easier to manage the code.

I adore WindowBuilder, but I do feel like this feature is very much needed and that it will be taken into consideration.

Example of what I mean:
**This is the current situation:**
`public class ExampleCurrent {

	private JFrame frame;
	private JButton btnNewButton;

	/**
	 * Launch the application.
	 */
	public static void main(String[] args) {
		EventQueue.invokeLater(new Runnable() {
			public void run() {
				try {
					ExampleCurrent window = new ExampleCurrent();
					window.frame.setVisible(true);
				} catch (Exception e) {
					e.printStackTrace();
				}
			}
		});
	}

	/**
	 * Create the application.
	 */
	public ExampleCurrent() {
		initialize();
	}

	/**
	 * Initialize the contents of the frame.
	 */
	private void initialize() {
		frame = new JFrame();
		frame.setBounds(100, 100, 450, 300);
		frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
		
		btnNewButton = new JButton("New button");
		
		JLabel lblNewLabel = new JLabel("New label");
		GroupLayout groupLayout = new GroupLayout(frame.getContentPane());
		groupLayout.setHorizontalGroup(
			groupLayout.createParallelGroup(Alignment.LEADING)
				.addGroup(groupLayout.createSequentialGroup()
					.addContainerGap()
					.addGroup(groupLayout.createParallelGroup(Alignment.LEADING)
						.addComponent(btnNewButton)
						.addComponent(lblNewLabel))
					.addContainerGap(335, Short.MAX_VALUE))
		);
		groupLayout.setVerticalGroup(
			groupLayout.createParallelGroup(Alignment.LEADING)
				.addGroup(groupLayout.createSequentialGroup()
					.addContainerGap()
					.addComponent(btnNewButton)
					.addGap(63)
					.addComponent(lblNewLabel)
					.addContainerGap(150, Short.MAX_VALUE))
		);
		frame.getContentPane().setLayout(groupLayout);
	}
}`

**This is what I would like this feature to do:**
`public class ExampleDesired {

	private JFrame frame;
	private JButton btnNewButton;
	private JLabel lblNewLabel;
	
	/**
	 * Launch the application.
	 */
	public static void main(String[] args) {
		EventQueue.invokeLater(new Runnable() {
			public void run() {
				try {
					ExampleDesired window = new ExampleDesired();
					window.frame.setVisible(true);
				} catch (Exception e) {
					e.printStackTrace();
				}
			}
		});
	}

	/**
	 * Create the application.
	 */
	public ExampleDesired() {
		initialize();
		initializeLayouts();
	}

	/**
	 * Initialize the contents of the frame.
	 */
	private void initialize() {
		frame = new JFrame();
		frame.setBounds(100, 100, 450, 300);
		frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
		
		btnNewButton = new JButton("New button");
		
		lblNewLabel = new JLabel("New label");
	}
	
	/**
	 * Initialize the layouts of the frame
	 */
	public void initializeLayouts() {
		GroupLayout groupLayout = new GroupLayout(frame.getContentPane());
		groupLayout.setHorizontalGroup(
			groupLayout.createParallelGroup(Alignment.LEADING)
				.addGroup(groupLayout.createSequentialGroup()
					.addContainerGap()
					.addGroup(groupLayout.createParallelGroup(Alignment.LEADING)
						.addComponent(btnNewButton)
						.addComponent(lblNewLabel))
					.addContainerGap(335, Short.MAX_VALUE))
		);
		groupLayout.setVerticalGroup(
			groupLayout.createParallelGroup(Alignment.LEADING)
				.addGroup(groupLayout.createSequentialGroup()
					.addContainerGap()
					.addComponent(btnNewButton)
					.addGap(63)
					.addComponent(lblNewLabel)
					.addContainerGap(150, Short.MAX_VALUE))
		);
		frame.getContentPane().setLayout(groupLayout);
	}
}`

**Comment by wimjongman** [comment]
I think you are looking for "lazy variable and statement generation"

![image](https://user-images.githubusercontent.com/180969/159117636-45d36d97-fd35-4ce5-bd2d-0bf1cdb93738.png)


Which delivers this code:

```

public class Test extends JPanel {
	private JLabel lblNewLabel;
	private JTextField textField;

	/**
	 * Create the panel.
	 */
	public Test() {
		setLayout(new GridLayout(0, 2, 0, 0));
		add(getLblNewLabel());
		add(getTextField());

	}

	private JLabel getLblNewLabel() {
		if (lblNewLabel == null) {
			lblNewLabel = new JLabel("New label");
		}
		return lblNewLabel;
	}
	private JTextField getTextField() {
		if (textField == null) {
			textField = new JTextField();
			textField.setColumns(10);
		}
		return textField;
	}
}

```

**Reply by MrRymen** [reply]
Thank you for the reply. Unfortunately, it is not what I meant. I would like the code generator to store anything layout related in a separate method, so that the rest of the code will be readable. My example used only one GroupLayout just for presentation purposes, and it already takes a lot of space. But in a real life project that has several tabs and more than 10 GroupLayouts in total, the initialize() method becomes very, very messy and difficult to read. Which is why I suggested adding a feature to store all layout code as a separate method while leaving all other components inside the initialize() same as by default.

**Comment by wimjongman** [comment]
<strike>This is the best we can do. It is already doing 90% of what you want.</strike> (EDIT: it seems that group layouts are not generated in this method)
 
Another option is to use a layout that is not so chatty. For example, GridLayout or [MigLayout](https://www.miglayout.com/).

<strike>I don't think you can persuade anyone to spend time implementing your specific requirement. If you want to spend your own time, then you are more than welcome to take a stab at this.</strike>

You are more than welcome to take a stab at this We will help you get started.

This would require the following:

1. [Setup a Windowbuilder development environment](https://github.com/eclipse/windowbuilder/blob/master/CONTRIBUTING.md) (like you mentioned ^^)
2. Make a new preference "Generate layout in a separate method"
3. Write generation code
4. Write parsing code
